02cpp1/sec03Pointers.md

```diff
@@ -0,0 +1,364 @@
+- Shared Ownership: The lifetime of the data is determined by multiple other variables, and the data should remain as long as one of those variables is still in scope. This is represented using **Shared Pointers**. Once all shared pointers pointing to a particular piece of data go out of scope, then the memory for that data is freed. 
+- Non-Owning: Non-owning pointers should have no impact on the lifetime of the data. When the non-owning pointer goes out of scope nothing happens to the data that it was pointing to. There are represented using **Weak Pointers** or traditional raw pointers. 
+
+## Unique Pointers `std::unique_ptr<>`
+
+A variable should be a *unique pointer* when that is the only variable that controls whether the resource should be destroyed or not. When a unique pointer goes out of scope, the object that it pointed to is deleted and the memory freed.
+
+```cpp
+#include <memory>
+#include <iostream>
+
+int main()
+{
+    std::unique_ptr<int> p_int = std::make_unique<int>(17);
+    std::unique_ptr<int> p_int2(new int(25))
+
+    std::cout << *p_int << std::endl;
+    std::cout << *p_int2 << std::endl;
+
+    return 0;
+}
+```
+- Remember that pointers actually store memory addresses, not the values of the variables that they point to. So to get the value of the variable we need to "dereference" the pointer using the `*` operator. `p_int` refers to the smart pointer, but `*p_int` gives us the value of the integer that we are pointing to. 
+- We can make assignments to `*p_int` which will update the value of the integer, but doesn't change the memory location (so `*p_int` will change, but `p_int` won't). 
+
+You also can't make a copy of a unique pointer, as then there would be a conflict over which one should handle the destruction of the object when it goes out of scope. This means that when we want to pass a unique pointer to a function, we cannot pass it by value, because this would involve making a copy. We can, however, pass a unique pointer by reference. 
+```cpp
+#include <memory>
+#include <iostream>
+
+void updatePtrValue(std::unique_ptr<int> &p)
+{
+    *p += 5;
+}
+
+int main()
+{
+    std::unique_ptr<int> p_int = std::make_unique<int>(17);
+    updatePtrValue(p_int);
+    std::cout << *p_int << std::endl;
+
+    return 0;
+}
+```
+This would usually be the preferred way to pass unique pointers to a function. If, for some reason, you don't want to pass by reference, then you will need to use `std::move`. 
+
+Since we can't have multiple unique pointers pointing to the same data, if we want to transfer the ownership of the data to a new unique pointer, we use `std::move` as follows:
+```cpp
+std::unique_ptr<int> p1 = std::make_unique<int>(5);
+std::unique_ptr<int> p2 = std::move(p1);
+```
+- **Avoid doing this if at all possible. After this operation, `p1` will no longer point to valid memory, and will cause a segmentation fault if accessed (i.e. your program will crash).** 
```

Comment:
Is there possibly a reference to a higher power we can use here? e.g. something from the [core guidelines](https://github.com/isocpp/CppCoreGuidelines)?

Reply:
I have softened this statement a bit to encourage people to think about the dangers of having pointers left empty like this.